isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java

```diff
@@ -89,22 +104,133 @@ public RelNode visit(NamedScan namedScan) throws RuntimeException {
+                gr ->
+                    gr.getExpressions().stream()
+                        .map(expr -> expr.accept(expressionRexConverter))
+                        .toList())
             .toList();
+    List<RexNode> groupExprs =
+        groupExprLists.stream().flatMap(Collection::stream).collect(Collectors.toList());
+    RelBuilder.GroupKey groupKey = relBuilder.groupKey(groupExprs, groupExprLists);
 
-    return relBuilder.push(child).project(rexList).build();
+    List<AggregateCall> aggregateCalls =
+        aggregate.getMeasures().stream().map(m -> fromMeasure(m)).toList();
+    return relBuilder.push(child).aggregate(groupKey, aggregateCalls).build();
+  }
+
+  private AggregateCall fromMeasure(Aggregate.Measure measure) {
+    var arguments =
+        measure.getFunction().arguments().stream()
+            .map(expr -> expr.accept(expressionRexConverter))
+            .toList();
+    var operator =
+        aggregateFunctionConverter.getSqlOperatorFromSubstraitFunc(
+            measure.getFunction().declaration().key(), measure.getFunction().outputType());
+    if (!operator.isPresent()) {
+      throw new UnsupportedOperationException(
+          String.format(
+              "Unable to find binding for call %s", measure.getFunction().declaration().name()));
+    }
+    List<Integer> argIndex = new ArrayList<>();
+    for (RexNode arg : arguments) {
+      if (!(arg instanceof RexInputRef)) {
+        throw new UnsupportedOperationException(
+            String.format(
+                "aggregation function only support column as argument. % is not support", arg));
+      }
+      argIndex.add(((RexInputRef) arg).getIndex());
+    }
+
+    boolean distinct =
+        measure.getFunction().invocation()
+            == AggregateFunction.AggregationInvocation.AGGREGATION_INVOCATION_DISTINCT;
+
+    SqlAggFunction aggFunction;
+    RelDataType returnType =
+        TypeConverter.convert(relOptCluster.getTypeFactory(), measure.getFunction().getType());
+
+    boolean approximate = false;
+    if (operator.get() == SqlStdOperatorTable.COUNT) {
+      aggFunction = SqlStdOperatorTable.COUNT;
+    } else if (operator.get() == SqlStdOperatorTable.SUM) {
+      aggFunction = SqlStdOperatorTable.SUM;
+    } else if (operator.get() == SqlStdOperatorTable.AVG) {
+      aggFunction = SqlStdOperatorTable.AVG;
+    } else if (operator.get() == SQL_OPERATOR_APPROX_COUNT_DISTINCT) {
+      aggFunction = SqlStdOperatorTable.COUNT;
+      approximate = true;
+      distinct = true;
+    } else {
+      throw new UnsupportedOperationException(
+          String.format(
+              "Unsupported agg operator % for agg function % ",
+              operator.get(), measure.getFunction().declaration().name()));
+    }
+
+    int filterArg = -1;
+    if (measure.getPreMeasureFilter().isPresent()) {
+      RexNode filter = measure.getPreMeasureFilter().get().accept(expressionRexConverter);
+      if (!(filter instanceof RexInputRef)) {
```

Comment:
Consider factoring this check out to a helper method as it's repeated above.

Reply:
Factor the common codes into a private method.